SwiftGit2/CommitIterator.swift

```diff
@@ -0,0 +1,67 @@
+//
+// Created by Arnon Keereena on 4/28/17.
+// Copyright (c) 2017 GitHub, Inc. All rights reserved.
+//
+
+import Result
+import libgit2
+
+public class CommitIterator: IteratorProtocol {
+    public typealias Element = Result<Commit, NSError>
+    var repo: Repository
+    var branch: Branch
+    var revisionWalker: OpaquePointer? = nil
+    var oid: git_oid
+    private var unsafeCommit: OpaquePointer? = nil
+    
+    public init(repo: Repository, branch: Branch) {
+        self.repo = repo
+        self.branch = branch
+        self.oid = branch.oid.oid
+        setupRevisionWalker()
+    }
+    
+    deinit {
+        git_revwalk_free(self.revisionWalker)
+    }
+    
+    private func setupRevisionWalker() {
+        git_revwalk_new(&revisionWalker, repo.pointer)
+        git_revwalk_sorting(revisionWalker, GIT_SORT_TOPOLOGICAL.rawValue)
+        git_revwalk_sorting(revisionWalker, GIT_SORT_TIME.rawValue)
+        git_revwalk_push(revisionWalker, &oid)
+    }
+    
+    private func result(withName name: String, from result: Int32) -> (stop: Bool, error: NSError?) {
+        guard result == GIT_OK.rawValue else {
+            if result == GIT_ITEROVER.rawValue {
+                return (stop: true, error: nil)
+            } else {
+                return (stop: false, error: NSError(gitError: result, pointOfFailure: name))
+            }
+        }
+        return (stop: false, error: nil)
```

Comment:
I think this would be clearer with a private enum.

```swift
public class CommitIterator: IteratorProtocol {
    …

    private enum Next {
        case over
        case ok
        case error(NSError)

        init(_ result: Int32, name: String) {
            switch result {
            case GIT_ITEROVER.rawValue:
                return .over
            case GIT_OK.rawValue:
                return .ok
            default:
                return .error(NSError(gitError: result, pointOfFailure: name))
            }
        }
    }
}
```